### PR TITLE
[IMPROVEMENT] added client retry for jetstream async publish old API

### PR DESCRIPTION
**Pull request by pranavmehta94**
This PR contains following:
1. Added retry logic for jestream async publish using old API
2. retries are controlled by retry wait and retry attempts publish options
3. no responder error is retried
4. pub ack future object is re-used in retries

This PR fixes following issues:

- https://github.com/nats-io/nats.go/issues/1678

### CODE REVIEW

File: js.go [hunk 1]
```diff
@@ -806,20 +812,23 @@ func (js *js) handleAsyncReply(m *Msg) {
...
-		// Defer here so error is processed and can be checked.
-		defer close(dch)
+
+	closeDch := func() {
+		// Check on anyone one waiting on done status.
+		if js.dch != nil && len(js.pafs) == 0 {
+			dch := js.dch
+			js.dch = nil
+			// Defer here so error is processed and can be checked.
+			defer close(dch)
```
Comment (piotrpio) [inline js.go:830]: This defer does not work as expected since it'll be executed at the end of `closeDch`, not the caller. While it would be good to extract this func since it's repeated twice, I do think that we should make sure that `close(dch)` is executed at the very end to first send signals on the `pubAckFuture` and only later to `PublishAsyncComplete()`

File: js.go [hunk 2]
```diff
@@ -894,7 +933,10 @@ func (js *js) PublishAsync(subj string, data []byte, opts ...PubOpt) (PubAckFutu
 const defaultStallWait = 200 * time.Millisecond
 
 func (js *js) PublishMsgAsync(m *Msg, opts ...PubOpt) (PubAckFuture, error) {
-	var o pubOpts
+	var o = pubOpts{
+		rwait: DefaultPubRetryWait,
```
Comment (piotrpio) [inline js.go:937]: Since this is an old API many people rely on, I would be hesitant to make the retries default, but rather treat them as opt-in and clearly state it in the doc. I know that this makes it work differently than the new API, but I don't think we should be changing the default behavior in this case.

File: js.go [hunk 3]
```diff
@@ -806,20 +812,23 @@ func (js *js) handleAsyncReply(m *Msg) {
...
+		// Check on anyone one waiting on done status.
+		if js.dch != nil && len(js.pafs) == 0 {
+			dch := js.dch
+			js.dch = nil
+			// Defer here so error is processed and can be checked.
+			defer close(dch)
+		}
 	}
 
 	doErr := func(err error) {
```
Comment (piotrpio) [inline js.go:834]: We should clear the reply subject before passing the message in error handler. If user wants to retry manually inside the error handler, having `paf.msg.Reply` set will cause the retry attempt to fail with `nats: reply subject should be empty`

File: object.go [hunk 4]
```diff
@@ -398,7 +398,7 @@ func (obs *obs) Put(meta *ObjectMeta, r io.Reader, opts ...ObjectOpt) (*ObjectIn
 		return nil
 	}
 
-	m, h := NewMsg(chunkSubj), sha256.New()
```
Comment (piotrpio) [inline object.go:401]: Why is this change necessary?

File: test/js_test.go
Comment (piotrpio) [inline test/js_test.go:1]: We should really have tests for async retries. For reference, see tests in the new API: https://github.com/nats-io/nats.go/blob/main/jetstream/test/publish_test.go#L1402

**Review by piotrpio — CHANGES_REQUESTED** [review]
Thanks for the contribution! I have a few comments.

File: object.go (hunk 4 above)
Comment (pranav-mehta) [reply object.go:401]: Earlier, reply field in msg was set to empty string on function exit using defer. After adding retry logic,  go test detected race between setting reply in async handler and emptying it in publish async msg. So, the defer call which emptied the reply inside msg was removed. 
The code was re-using the msg object inside the loop which called publish async msg and the test failed complaining about nats reply being not empty. Hence, the new msg object is being created in every iteration of the loop now

**Comment by pranav-mehta** [comment]
@piotrpio I incorporated the review comments but the CI pipeline now fails in an unrelated test. When I run the tests locally, all the test succeedd

File: object.go (hunk 4 above)
Comment (pranavmehta94) [reply object.go:401]: Earlier, reply field in msg was set to empty string on `PublishMsgAsync` exit using defer. After adding retry logic, go test detected race between setting reply in async handler `handleAsyncReply` and emptying it in `PublishMsgAsync`. So, the defer call which emptied the reply inside PublishMsgAsync was removed. The code was re-using the msg object inside the loop which called `PublishMsgAsync` and the test failed complaining about nats reply being not empty. Hence, the new msg object is being created in every iteration of the loop now

**Comment by pranavmehta94** [comment]
[@piotrpio](https://github.com/piotrpio) I incorporated the review comments but the CI pipeline now fails in an unrelated test. When I run the tests locally, all the tests succeed

File: object.go (hunk 4 above)
Comment (piotrpio) [reply object.go:401]: This is no good I think. We should not leave `Msg.Reply` set with the generated subject as it makes it impossible to re-publish by the user (as in the object store now). By the way, having `Put()` in object store recreate the message with each iteration means we have increased allocations (especially for large payloads / small chunks).

I may have an idea of how to tackle this problem - we could stop modifying the message passed by the user (which would prevent race) and utilize the low level `nc.publish()`. Here's my initial attempt from your branch: https://gist.github.com/piotrpio/796e5203e2d5be2cabda771e9ec51d54

Possibly this can be improved, but it seems to work.

Comment (piotrpio) [reply object.go:401]: Ok but even so, with your changes the message will never be restored to its original state right? Even after the message has been processed. If we need to fix it anyway we can also improve by not modifying the message at all I think.

Comment (pranavmehta94) [reply object.go:401]: I understand that a caller would assume that message would not be modified in anyway after  `PublishMsgAsync ` is called and `PubAckFuture ` is processed. This is the reason why Reply was not unset earlier. I will update the PR by applying your patch

Comment (piotrpio) [reply object.go:401]: But it was unset before your changes: https://github.com/nats-io/nats.go/pull/1695/files#diff-176a2342c55218b70f165bed6a7731dc83d901f9513382f3c98bbd3d08ce0678L941

After `PubAckFuture` is processed, the message can be re-used by the user and thus should be in its original state:
`/ The message should not be changed *until* the PubAckFuture has been processed.`

Again, your changes are good in general and we should definitely process those but we also need to be very careful when changing any behavior in the old JetStream API.

Comment (pranavmehta94) [reply object.go]: Yes I understand it was unset before these changes as well

Comment (pranavmehta94) [reply object.go]: @piotrpio I have updated the PR by applying your patch with one small patch. Please review the changes

**Review by piotrpio — APPROVED** [review]
LGTM, thanks for the contribution!
